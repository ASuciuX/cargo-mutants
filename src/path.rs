use camino::Utf8Path;

pub trait Utf8PathSlashes {
    fn to_slash_path(&self) -> String;
}

impl Utf8PathSlashes for Utf8Path {
    fn to_slash_path(&self) -> String {
        self.components()
            .map(|c| c.as_str())
            .collect::<Vec<_>>()
            .join("/")
    }
}
