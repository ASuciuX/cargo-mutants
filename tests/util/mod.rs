// Copyright 2024 Martin Pool

#![allow(dead_code)] // rustc doesn't understand they're used by multiple crates?

//! Reusable utilities for cargo-mutants tests.

use std::borrow::Borrow;
use std::env;
use std::path::{Path, PathBuf};
use std::time::Duration;

use lazy_static::lazy_static;
use tempfile::{tempdir, TempDir};

/// A timeout for a `cargo mutants` invocation from the test suite. Needs to be
/// long enough that even commands that do a lot of work can pass even on slow
/// CI VMs and even on Windows, but short enough that the test does not hang
/// forever.
pub const OUTER_TIMEOUT: Duration = Duration::from_secs(60);

lazy_static! {
    pub static ref MAIN_BINARY: PathBuf = assert_cmd::cargo::cargo_bin("cargo-mutants");
}

pub fn run() -> assert_cmd::Command {
    let mut cmd = assert_cmd::Command::new(MAIN_BINARY.as_os_str());
    // Strip any options configured in the environment running these tests,
    // so that they don't cause unexpected behavior in the code under test.
    //
    // For example, without this,
    // `env CARGO_MUTANTS_JOBS=4 cargo mutants`
    //
    // would end up with tests running 4 jobs by default, which would cause
    // the tests to fail.
    //
    // Even more generally than that example, we want the tests to be as hermetic
    // as reasonably possible.
    env::vars()
        .map(|(k, _v)| k)
        .filter(|k| {
            k.starts_with("CARGO_MUTANTS_")
                || k == "CLICOLOR_FORCE"
                || k == "NOCOLOR"
                || k == "CARGO_TERM_COLOR"
        })
        .for_each(|k| {
            cmd.env_remove(k);
        });
    cmd
}

pub trait CommandInstaExt {
    fn assert_insta(&mut self, snapshot_name: &str);
}

impl CommandInstaExt for assert_cmd::Command {
    fn assert_insta(&mut self, snapshot_name: &str) {
        let output = self.output().expect("command completes");
        assert!(output.status.success());
        insta::assert_snapshot!(snapshot_name, String::from_utf8_lossy(&output.stdout));
        assert_eq!(&String::from_utf8_lossy(&output.stderr), "");
    }
}

// Copy the source because output is written into mutants.out.
pub fn copy_of_testdata(tree_name: &str) -> TempDir {
    let tmp_src_dir = tempdir().unwrap();
    cp_r::CopyOptions::new()
        .filter(|path, _stat| {
            Ok(["target", "mutants.out", "mutants.out.old"]
                .iter()
                .all(|p| !path.starts_with(p)))
        })
        .copy_tree(Path::new("testdata").join(tree_name), &tmp_src_dir)
        .unwrap();
    tmp_src_dir
}

/// Assert that some bytes, when parsed as json, equal a json value.
pub fn assert_bytes_eq_json<J: Borrow<serde_json::Value>>(actual: &[u8], expected: J) {
    // The Borrow is so that you can pass either a value or a reference, for easier
    // calling.
    let actual_json = std::str::from_utf8(actual)
        .expect("bytes are UTF-8")
        .parse::<serde_json::Value>()
        .expect("bytes can be parsed as JSON");
    assert_eq!(&actual_json, expected.borrow());
}
