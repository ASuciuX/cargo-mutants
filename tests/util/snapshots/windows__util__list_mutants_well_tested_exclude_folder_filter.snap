---
source: tests/util/mod.rs
expression: "String::from_utf8_lossy(&output.stdout)"
---
src/methods.rs:2:5: replace double -> usize with 0
src/methods.rs:2:5: replace double -> usize with 1
src/methods.rs:2:7: replace * with + in double
src/methods.rs:2:7: replace * with / in double
src/module/module_methods.rs:2:5: replace double -> usize with 0
src/module/module_methods.rs:2:5: replace double -> usize with 1
src/module/module_methods.rs:2:7: replace * with + in double
src/module/module_methods.rs:2:7: replace * with / in double
src/module/utils/inside_mod.rs:4:13: replace outer::inner::name -> &'static str with ""
src/module/utils/inside_mod.rs:4:13: replace outer::inner::name -> &'static str with "xyzzy"
src/module/utils/nested_function.rs:2:5: replace has_nested -> u32 with 0
src/module/utils/nested_function.rs:2:5: replace has_nested -> u32 with 1
src/module/utils/nested_function.rs:3:9: replace has_nested::inner -> u32 with 0
src/module/utils/nested_function.rs:3:9: replace has_nested::inner -> u32 with 1
src/module/utils/nested_function.rs:5:13: replace * with + in has_nested
src/module/utils/nested_function.rs:5:13: replace * with / in has_nested
src/module/utils/sub_utils/subutils_nested_function.rs:2:5: replace has_nested -> u32 with 0
src/module/utils/sub_utils/subutils_nested_function.rs:2:5: replace has_nested -> u32 with 1
src/module/utils/sub_utils/subutils_nested_function.rs:3:9: replace has_nested::inner -> u32 with 0
src/module/utils/sub_utils/subutils_nested_function.rs:3:9: replace has_nested::inner -> u32 with 1
src/module/utils/sub_utils/subutils_nested_function.rs:5:13: replace * with + in has_nested
src/module/utils/sub_utils/subutils_nested_function.rs:5:13: replace * with / in has_nested
